### DISCUSSION TITLE
Swap out youtube-dl for yt-dlp ?

### DISCUSSION THREAD

**Discussion by Russsgithub** [savonet/liquidsoap#2832] [Ideas]
I've recently been playing with youtube videos and noticed that youtube-dl is being very slow at downloading the videos. @smimram increased the timeout yesterday, which fixes the cmd failing from timeout issue.

After having a hunt around I found that this is a possible [issue](https://github.com/ytdl-org/youtube-dl/issues/30102) with youtube-dl which users have been fixing by using yt-dlp ( a [youtube-dl](https://github.com/ytdl-org/youtube-dl) fork based on the now inactive [youtube-dlc](https://github.com/blackjack4494/yt-dlc) )

I wrote a protocol for yt-dlp ( needs [yt-dlp](https://github.com/yt-dlp/yt-dlp) installed ) based on the youtube-dl protocol.

```

let settings.protocol.yt_dlp = settings.make.protocol("yt-dlp")

let settings.protocol.yt_dlp.path = settings.make(
  description="Path of the yt-dlp binary",
  "yt-dlp"
)

let settings.protocol.yt_dlp.timeout = settings.make(
  description="Timeout (in seconds) for yt-dlp executions.",
  300.
)

# Register the youtube-dl protocol, using youtube-dl.
# Syntax: youtube-dl:<ID>
# @flag hidden
def yt_dlp_protocol(~rlog,~maxtime,arg)
  binary = settings.protocol.yt_dlp.path()
  timeout = settings.protocol.yt_dlp.timeout()
  height = settings.frame.video.height()

  def log(~level,s) =
    rlog(s)
    log(label="protocol.yt_dlp",level=level,s)
  end

  delay = maxtime - time()

  ### Added -format-sort arg which picks the best video matching or less than the video frame height set in liquidsoap settings.

  cmd = "#{binary} --get-title --get-filename --format-sort res:#{height} -- #{process.quote(arg)}"
  log(level=4,"Executing #{cmd}")
  x = process.read.lines(timeout=delay,cmd)

  x =
    if list.length(x) >= 2 then
      x
    else
      ["",".osb"]
    end

  title = list.hd(default="",x)
  ext   = file.extension(leading_dot=false,list.nth(default="",x,1))

  cmd = "rm -f $(output) && #{binary} -q -f best --no-playlist -o $(output) -- #{process.quote(arg)}"
  cmd = process_uri(extname=ext, cmd)

  if title != "" then
    ["annotate:title=#{string.quote(title)}:#{cmd}"]
  else
    [cmd]
  end
end
add_protocol("yt-dlp", yt_dlp_protocol,
              doc="Resolve a request using yt-dlp.",
              syntax="yt-dlp:uri")
              
 ```
 
 The above protocol adds ```-format-sort res:#{height}``` which limits the the resolution of the downloaded video to the best video available that matches or is under the video frame height set in liquidsoap ```settings.frame.video.height()```
 
 The protocol is then called with ```yt-dlp:uri```

So thought I'd share.

This massively improves the download speeds from youtube , uses the same syntax as youtube-dl ( as far as I can see ) and seems to have many additional features which I'd be happy to work some of into the protocol if that would be useful as a pull request.

It can also be used by replacing the youtube-dl path with the yt-dlp path in scripts using

```settings.protocol.youtube_dl.path.set("yt-dlp")``` 

with  [yt-dlp](https://github.com/yt-dlp/yt-dlp) installed which change the standard youtube-dl protocol to use yt-dlp instead of youtube-dl but may affect the usage of youtube-pl protocol (I have not tested this) and does not included the resolution limiting feature of yt-dlp.

Notes.......

The protocol was written for liquidsoap version 2.1.3 ( I'm running stable release at the moment) in 2.2 I believe the syntax of ```add_protocol``` changes to ```protocol.add```

**Comment by toots** [comment]
Hi @Russsgithub and thanks for the detailed analysis and contribution. I sure wish we'd stop forking projects all the time haha..

Have you tried to simply change the [path to `youtube-dl`](https://www.liquidsoap.info/doc-2.1.3/settings.html#path-of-the-youtube-dl-binary):
```
settings.protocol.youtube_dl.path.set("yt-dlp ")
```
If the options we rely on are similar enough, this should work just fine..

**Comment by Russsgithub** [comment]
@tootsI'm sure that would work.Just thought I'd share findings as youtube-dll is being really slow for whatever reason.I'm unsure how just changing the binary path would affect the youtube-pl protocol.I do like the resolution limiting function in yt-dlp which would prevent downloading say 1080p files to be broadcast as say 480p.I get you on forking projects.Strangely it seems the code from yt-dlp seems at least in part to be being merged into youtube-dl. I've not done any tests but there are reports of around 100kb/s in one and 10mb/s in the other.ThanksOn 6 Jan 2023 14:25, Romain Beauxis ***@***.***> wrote:
Hi @Russsgithub and thanks for the detailed analysis and contribution. I sure wish we'd stop forking projects all the time haha..
Have you tried to simply change the path to youtube-dl:
settings.protocol.youtube_dl.path.set("yt-dlp ")

If the options we rely on are similar enough, this should work just fine..
